peer-file-watcher.js

```diff
@@ -0,0 +1,197 @@
+
+    // Trigger an initial load
+    this.reloadSync();
+
+    function _boundReload() {
+        self.reload();
+    }
+}
+
+PeerFileWatcher.prototype._validateFilePath = function _validateFilePath() {
+    if (!fs.existsSync(this.peerFile)) {
+        assert(false, 'Peer hosts file does not exist: ' + this.peerFile);
+    }
+};
+
+PeerFileWatcher.prototype._readPeerListFromFileSync =
+function _readPeerListFromFileSync() {
+    var tuple = safeJSONParse(fs.readFileSync(this.peerFile, 'utf8'));
+    if (tuple[0]) {
+        assert(false, 'Invalid JSON for TChannel peer host file at ' + this.peerFile);
+    }
+    return tuple[1];
+};
+
+PeerFileWatcher.prototype._establishFileWatcher =
+function _establishFileWatcher() {
+    // Use watchFile instead of watch here. watch uses inotify events,
+    // but since we're getting renames every 30 seconds and the inode is
+    // changing we'd need to keep creating new watches for every file and
+    // eventually exhaust the kernel's inotify watch limit.
+    //
+    // watchFile instead polls with stat. Not as performant as inotify, but
+    // better than watches eventually failing.
+    fs.watchFile(this.peerFile, {
+        interval: this.refreshInterval,
+        persistent: true
+    }, this._boundReload);
+};
+
+PeerFileWatcher.prototype.destroy = function destroy() {
+    fs.unwatchFile(this.peerFile, this._boundReload);
+};
+
+/**
+ * Load hosts from a JSON file and update peers in TChannel.
+ */
+PeerFileWatcher.prototype.reloadSync =
+function reloadSync() {
+    this.logger.info('PeerFileWatcher: Loading peer list from file sync', {
+        peerFile: this.peerFile
+    });
+
+    var newPeers = this._readPeerListFromFileSync();
+    this.updatePeers(newPeers);
+};
+
+PeerFileWatcher.prototype.reload =
+function reload() {
+    var self = this;
+    this.logger.info('PeerFileWatcher: Loading peer list from file async', {
+        peerFile: this.peerFile
+    });
+
+    this._readPeerList(onPeers);
+
+    function onPeers(err, newPeers) {
+        if (err) {
+            self.logger.error('PeerFileWatcher: Could not load peers file', {
+                peerFile: self.peerFile,
+                error: err
+            });
+            return;
+        }
+
+        self.updatePeers(newPeers);
+    }
+};
+
+PeerFileWatcher.prototype._readPeerList =
+function _readPeerList(cb) {
+    fs.readFile(this.peerFile, 'utf8', onFile);
+
+    function onFile(err, text) {
+        if (err) {
+            return cb(err);
+        }
+
+        var tuple = safeJSONParse(text);
+        if (tuple[0]) {
+            return cb(tuple[0]);
+        }
+
+        cb(null, tuple[1]);
+    }
+};
+
+PeerFileWatcher.prototype.updatePeers =
+function updatePeers(newPeers) {
+    // Take a snapshot of current, existing peers. This is used to delete old
+    // peers later.
+    var oldPeers = this.channel.peers.keys().slice();
+
+    var i;
+
+    // Load new peers; duplicates are ignored
```

Comment:
just to double check, is the comment accurate? where are the duplicates ignored?